### PR TITLE
Improve jsre

### PR DESCRIPTION
**Pull request by juancarlospaco**
- Add dollar `$` for `jsre.RegExp`.

- Unrelated CI fail:
```
[apt-fast 15:37:55] Working... this may take a while.
E: Unable to correct problems, you have held broken packages.
[apt-fast 15:37:55] Package manager quit with exit code.
```

## future work
- [ ] https://github.com/nim-lang/Nim/pull/17365/files/f3b41718f6c3d72bd95efb8c2bd6c5d2ddee320d..4e32ba179f48c67f58b39344ebb8cf87b25a464d#r593792475

> how about making the API more complete by adding wrappers for `matchAll` + friends:

> These patterns are used with the exec() and test() methods of RegExp, and with the match(), matchAll(), replace(), replaceAll(), search(), and split() methods of String.

(https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Regular_Expressions)

- [ ] https://github.com/nim-lang/Nim/pull/17365#discussion_r597972679 `lastMatch` + friends don't work (static fields in js)

### CODE REVIEW

File: lib/js/jsre.nim [hunk 1]
```diff
@@ -4,12 +4,13 @@
 runnableExamples:
   let jsregex: RegExp = newRegExp(r"\s+", r"i")
   jsregex.compile(r"\w+", r"i")
-  doAssert jsregex.test(r"nim javascript")
-  doAssert jsregex.exec(r"nim javascript") == @["nim".cstring]
-  doAssert jsregex.toString() == r"/\w+/i"
+  assert jsregex.test(r"nim javascript")
```
Comment (timotheecour) [inline lib/js/jsre.nim]: how about introduce `match` and deprecate `test` as a deprecated alias so that it fits better with nim's various regex modules (re, nre, pkg/regex) ?
(sorry i didn't spot this in https://github.com/nim-lang/Nim/pull/14870)
should be done in a followup PR though to avoid mixing cleanups and API changes

Comment (timotheecour) [inline lib/js/jsre.nim]: `test` docs should mention whether it's a full match or substring match

File: lib/js/jsre.nim [hunk 2]
```diff
@@ -44,3 +45,7 @@ func test*(self: RegExp; pattern: cstring): bool {.importjs: "#.test(#)".}
 
 func toString*(self: RegExp): cstring {.importjs: "#.toString()".}
```
Comment (timotheecour) [inline lib/js/jsre.nim]: this should be deprecated as mentioned by @araq in various PR's eg: https://github.com/nim-lang/Nim/pull/16409#discussion_r549346583

* add a `toCstring`
* deprecate this one and make sure it's not used in docs/tests

File: lib/js/jsre.nim [hunk 3]
```diff
@@ -44,3 +45,7 @@ func test*(self: RegExp; pattern: cstring): bool {.importjs: "#.test(#)".}
...
   ## Returns a string representing the RegExp object.
+
+func `$`*(self: RegExp): string = $toString(self)
+
+func len*(self: RegExp): int = self.source.len
```
Comment (timotheecour) [inline lib/js/jsre.nim]: * i don't think we need this, since its code can be inlined and there's also not a javascript API for this; the RegExp object contains many fields, it's unclear why special case the length of the source field; but it's ok to add a line in runnableExamples mentioning:
`assert jsregex.source.len == 5`

* OTOH, `$` is fine I guess

File: lib/js/jsre.nim [hunk 4]
```diff
@@ -44,3 +45,7 @@ func test*(self: RegExp; pattern: cstring): bool {.importjs: "#.test(#)".}
 
 func toString*(self: RegExp): cstring {.importjs: "#.toString()".}
   ## Returns a string representing the RegExp object.
+
```
Comment (timotheecour) [inline lib/js/jsre.nim:38]: how about making the API more complete by adding wrappers for `matchAll` + friends:

> These patterns are used with the exec() and test() methods of RegExp, and with the match(), matchAll(), replace(), replaceAll(), search(), and split() methods of String.

(https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Regular_Expressions)

can be done in future PR of course.

Comment (juancarlospaco) [reply lib/js/jsre.nim:38]: Ye

File: lib/js/jsre.nim [hunk 5]
```diff
@@ -1,21 +1,9 @@
...
-  assert jsregex.test(r"0123456789abcd")
-  assert $jsregex == "/[0-9]/i"
-  assert jsregex.len == 5
-
 when not defined(js):
   {.error: "This module only works on the JavaScript platform".}
 
-type RegExp* {.importjs.} = object    ## Regular Expressions for JavaScript target.
+type RegExp* {.importjs.} = ref object of JsRoot ## Regular Expressions for JavaScript target.
   flags* {.importjs.}: cstring        ## cstring that contains the flags of the RegExp object.
```
Comment (timotheecour) [inline lib/js/jsre.nim]: please double check: are these `{.importjs.}` needed given that type itself is importjs?

File: lib/js/jsre.nim [hunk 6]
```diff
@@ -40,12 +28,27 @@ func compile*(self: RegExp; pattern: cstring; flags: cstring) {.importjs: "#.com
...
 
-func toString*(self: RegExp): cstring {.importjs: "#.toString()".}
+func toCstring*(self: RegExp): cstring {.importjs: "#.toString()".}
   ## Returns a string representing the RegExp object.
 
-func `$`*(self: RegExp): string = $toString(self)
+func `$`*(self: RegExp): string = $toCstring(self)
+
+func test*(self: RegExp; pattern: cstring): bool {.importjs: "#.test(#)", deprecated: "Use match instead".}
+  ## .. warning:: Deprecated, do not use.
```
Comment (timotheecour) [inline lib/js/jsre.nim]:
```suggestion
```

because https://github.com/nim-lang/Nim/pull/17054 will be merged soon, hopefully

File: lib/js/jsre.nim [hunk 7]
```diff
@@ -40,12 +28,27 @@ func compile*(self: RegExp; pattern: cstring; flags: cstring) {.importjs: "#.com
...
 
-func len*(self: RegExp): int = self.source.len
+func toString*(self: RegExp): cstring {.importjs: "#.toString()", deprecated: "Use toCstring instead".}
+  ## .. warning:: Deprecated, do not use.
```
Comment (timotheecour) [inline lib/js/jsre.nim]:
```suggestion
```
ditto

File: lib/js/jsre.nim [hunk 8]
```diff
@@ -1,20 +1,9 @@
...
-  doAssert jsregex.toString() == r"/\w+/i"
-  jsregex.compile(r"[0-9]", r"i")
-  doAssert jsregex.test(r"0123456789abcd")
-
-
 when not defined(js):
   {.error: "This module only works on the JavaScript platform".}
 
-type RegExp* {.importjs.} = object    ## Regular Expressions for JavaScript target.
+type RegExp* {.importjs.} = ref object of JsRoot ## Regular Expressions for JavaScript target.
```
Comment (timotheecour) [inline lib/js/jsre.nim]:
```suggestion
type RegExp* {.importjs.} = ref object of JsRoot
  ## Regular Expressions for JavaScript target.
  ## See https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/RegExp
```

Comment (konsumlamm) [reply lib/js/jsre.nim]:
```suggestion
type RegExp* {.importjs.} = ref object of JsRoot
  ## Regular Expressions for the JavaScript target.
  ## See https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/RegExp.
```
This is better imo.

File: lib/js/jsre.nim [hunk 9]
```diff
@@ -39,8 +28,27 @@ func compile*(self: RegExp; pattern: cstring; flags: cstring) {.importjs: "#.com
 func exec*(self: RegExp; pattern: cstring): seq[cstring] {.importjs: "#.exec(#)".}
   ## Executes a search for a match in its string parameter.
 
-func test*(self: RegExp; pattern: cstring): bool {.importjs: "#.test(#)".}
-  ## Tests for a match in its string parameter.
+func match*(self: RegExp; pattern: cstring): bool {.importjs: "#.test(#)".}
```
Comment (timotheecour) [inline lib/js/jsre.nim]: > func newRegExp*(pattern: cstring; flags: cstring): RegExp {.importjs: "new RegExp(@)".}

`flags` is optional, see https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/RegExp/RegExp

File: lib/js/jsre.nim [hunk 10]
```diff
@@ -3,27 +3,30 @@
...
+  lastParen*: cstring    ## Ditto.
+  leftContext*: cstring  ## Ditto.
+  rightContext*: cstring ## Ditto.
+
 
 func newRegExp*(pattern: cstring; flags: cstring): RegExp {.importjs: "new RegExp(@)".}
   ## Creates a new RegExp object.
 
+func newRegExp*(pattern: cstring): RegExp {.importjs: "new RegExp(@)".}
+
```
Comment (timotheecour) [inline lib/js/jsre.nim:29]: now that i realize that `test` is a substring match, how about naming it contains instead of match, and changing the order of arguments:

rationale:
* in pkg/regex, https://nitely.github.io/nim-regex/regex.html#match%2Cstring%2CRegex%2CRegexMatch%2Cint uses `contains` for this: `func contains(s: string; pattern: Regex): bool`
* in pkg/regex, `match` indicates a full match, and `find` is for a substring match (but also takes extra params: `func find(s: string; pattern: Regex; m: var RegexMatch; start = 0): bool`
* ditto with std/re: https://nim-lang.github.io/Nim/re.html#contains%2Cstring%2CRegex%2Cint
* ditto with std/nre: https://nim-lang.github.io/Nim/nre.html#contains%2Cstring%2CRegex%2Cint

basically, remove `match` and add this:
```nim
func contains*(pattern: cstring, self: RegExp): bool =
  ## Tests for a substring match in its string parameter.
  runnableExamples:
    let r = newRegExp(r"bc$", r"i")
    assert r in r"abc"
    assert r notin r"abcd"
    assert "xabc".contains(r)
  asm "`result` = `self`.test(`pattern`)"
```

File: lib/js/jsre.nim [hunk 11]
```diff
@@ -3,27 +3,30 @@
...
+  dotAll*: bool          ## Whether `.` matches newlines or not.
+  global*: bool          ## Whether to test against all possible matches in a string, or only against the first.
+  ignoreCase*: bool      ## Whether to ignore case while attempting a match in a string.
+  multiline*: bool       ## Whether to search in strings across multiple lines.
+  source*: cstring       ## The text of the pattern.
+  sticky*: bool          ## Whether the search is sticky.
+  unicode*: bool         ## Whether Unicode features are enabled.
+  lastIndex*: cint       ## Index at which to start the next match (read/write property).
+  input*: cstring        ## Read-only and modified on successful match.
+  lastMatch*: cstring    ## Ditto.
```
Comment (timotheecour) [inline lib/js/jsre.nim:19]: @juancarlospaco let's handle this in a followup PR since it's unrelated to the other changes and there is a choice to make

------------------------------

at least this field is incorrect for 2 reasons:
* first, it should have a warning: non-standard (refs: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/RegExp/lastMatch)
* second, and most importantly, it's a static property, as explained in the docs and by this example:

```nim
import std/jsre
let re = newRegExp "hi"
assert re.test("hi there")
echo re.lastMatch # TypeError: Cannot read property 'length' of undefined
```

## proposal 1
```nim
let regExp* {.importc: "RegExp".}: RegExp
runnableExamples:
  let re = newRegExp "hi"
  assert re.test("hi there") # (but use the new un-deprecated name)
  assert regExp.lastMatch == "hi"
```

## proposal 2
```nim
proc lastMatch*(T: typedesc[RegExp]): cstring {.importjs: "RegExp.lastMatch@".}
  # proc, not func, since there's a side effect here
runnableExamples:
  let re = newRegExp "hi"
  assert re.test("hi there") # (but use the new un-deprecated name)
  assert RegExp.lastMatch() == "hi"
```

## note
we need to check for all the other fields whether they're static or not

File: lib/js/jsre.nim [hunk 12]
```diff
@@ -45,13 +42,22 @@ func test*(self: RegExp; pattern: cstring): bool {.importjs: "#.test(#)", deprec
 
 func toString*(self: RegExp): cstring {.importjs: "#.toString()", deprecated: "Use toCstring instead".}
 
+func contains*(pattern: cstring; self: RegExp): bool =
```
Comment (timotheecour) [inline lib/js/jsre.nim:45]: func test*(self: RegExp; pattern: cstring): bool {.importjs: "#.test(#)", deprecated: "Use match instead".}

=>

func test*(self: RegExp; pattern: cstring): bool {.importjs: "#.test(#)", deprecated: "Use contains instead".}

**Review by timotheecour — APPROVED** [review]
LGTM after last nit(deprecation message)

File: lib/js/jsre.nim (hunk 11 above)
Comment (juancarlospaco) [reply lib/js/jsre.nim:19]: Does not work for me, `RegExp.lastMatch()` says `Error: object constructor needs an object type`.

**Review by timotheecour — APPROVED** [review]
